---
title: "GettingStarted"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


Importing TCGA MAF files for liver, testicular, and Lymphoid Neoplasm Diffuse Large B-cell Lymphoma.
```{r libraries, include=FALSE}
library(TCGAbiolinks)
thca_maf <- GDCquery_Maf("THCA", pipelines = "mutect")
dlbc_maf <- GDCquery_Maf("DLBC", pipelines = "mutect")
```

Extracting variant information from MAF
```{r variant extraction}
thca.variants <- extract_variants_from_matrix(thca_maf,
                                              chromosome_col="Chromosome",
                                              end_col="End_Position",
                                              start_col="Start_Position",
                                              ref_col="Tumor_Seq_Allele1",
                                              alt_col="Tumor_Seq_Allele2",
                                              sample_col="Tumor_Sample_Barcode")
lymphoma.variants <- extract_variants_from_matrix(dlbc_maf, 
                                              chromosome_col="Chromosome", 
                                              end_col="End_Position", 
                                              start_col="Start_Position", 
                                              ref_col="Tumor_Seq_Allele1", 
                                              alt_col="Tumor_Seq_Allele2",
                                              sample_col="Tumor_Sample_Barcode")
g <- select_genome("hg38")

```
Choose genome and create musica object with reference genome hg38
```{r build musica}
thca.musica <- create_musica(x = thca.variants, genome=g)
dlbc.musica <- create_musica(x = tgct.variants, genome=g)
```
Create SBS96 mutation table
```{r create mutation tables}

build_standard_table(thca.musica, g = g, table_name = "SBS96")
build_standard_table(dlbc.musica, g=g, table_name="SBS96")
```

Discover signatures for liver, testi, and lymphoma. Number of signatures were estimated using known cosmic signature numbers.
```{r discover signatures / exposures}

thca.result <- discover_signatures(musica = thca.musica, table_name = "SBS96",
                                   num_signatures = 4, method = "lda")
dlbc.result <- discover_signatures(musica = dlbc.musica, table_name = "SBS96", 
                              num_signatures = 6, method = "lda")
```
Plotting musica signatures
```{r plot musica}
plot_signatures(thca.result)
plot_exposures(thca.result, proportional = F)
plot_exposures(thca.result, proportional = TRUE)
plot_signatures(prad.result)
plot_exposures(prad.result, proportional = F)
plot_exposures(prad.result, proportional = TRUE)
plot_signatures(dlbc.result)
plot_exposures(dlbc.result, proportional = F)
plot_exposures(dlbc.result, proportional = TRUE)

```
Compare our signature findings to COSMIC signatures.
```{r exposure}
data("cosmic_v2_sigs")
 cosmic_v2_subtype_map("thyroid")
thca.pred_cosmic <- predict_exposure(musica = thca.musica, table_name = "SBS96",
                               signature_res = cosmic_v2_sigs,
                               signatures_to_use =  c(1, 2, 5, 13),
                               algorithm = "lda")
cosmic_v2_subtype_map("lymphoma")
dlbc.pred_cosmic <- predict_exposure(musica = dlbc.musica, table_name = "SBS96",
                               signature_res = cosmic_v2_sigs,
                               signatures_to_use =  c(1,  2,  5,  9, 13, 17),
                               algorithm = "lda")

thca.pred_our_sigs <- predict_exposure(musica = thca.musica, table_name = "SBS96",
                                 signature_res = thca.result, algorithm = "lda")
dlbc.pred_our_sigs <- predict_exposure(musica = dlbc.musica, table_name = "SBS96",
                                 signature_res = dlbc.result, algorithm = "lda")
compare_results(result = thca.pred_cosmic, other_result = thca.pred_our_sigs, threshold = 0.7)
compare_results(result = dlbc.pred_cosmic, other_result = dlbc.pred_our_sigs, threshold = 0.75)

```



