---
title: "GettingStarted"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


Importing the files
```{r cars}
library(TCGAbiolinks)
lihc_maf <- GDCquery_Maf("LIHC", pipelines = "mutect")
tgct_maf <- GDCquery_Maf("TGCT", pipelines = "mutect")
```

Extracting variant information from maf
```{r variant extraction}
lihc.variants <- extract_variants_from_matrix(lihc_maf, 
                                              chromosome_col="Chromosome", 
                                              end_col="End_Position", 
                                              start_col="Start_Position", 
                                              ref_col="Tumor_Seq_Allele1", 
                                              alt_col="Tumor_Seq_Allele2", 
                                              sample_col="Tumor_Sample_Barcode")
tgct.variants <- extract_variants_from_matrix(tgct_maf, 
                                              chromosome_col="Chromosome", 
                                              end_col="End_Position", 
                                              start_col="Start_Position", 
                                              ref_col="Tumor_Seq_Allele1", 
                                              alt_col="Tumor_Seq_Allele2", 
                                              sample_col="Tumor_Sample_Barcode")
```
Choose genome and create musica object
```{r build musica}
g <- select_genome("hg38")
lihc.musica <- create_musica(x = lihc.variants, genome = g)
tgct.musica <- create_musica(x = tgct.variants, genome=g)
build_standard_table(lihc.musica, g = g, table_name = "SBS96")
build_standard_table(tgct.musica, g=g, table_name="SBS96")

lihc.result <- discover_signatures(musica = lihc.musica, table_name = "SBS96", 
                              num_signatures = 4, method = "lda", nstart = 10)
tgct.result <- discover_signatures(musica = tgct.musica, table_name = "SBS96", 
                              num_signatures = 4, method = "lda", nstart = 10)
```
Plotting signatures
```{r plot musica}
plot_signatures(lihc.result)
plot_signatures(tgct.result)

plot_exposures(lihc.result, proportional = TRUE)
plot_exposures(tgct.result, proportional = TRUE)

lihc.samples <- get_sample_names(lihc.musica)
#plot_sample_counts(lihc.musica, sample_name = lihc.samples[c(3,4,5)], table_name = "SBS96")
tgct.samples <- get_sample_names(tgct.musica)
#plot_sample_counts(tgct.musica, sample_name = tgct.samples[c(3,4,5)], table_name = "SBS96")

```
```{r exposure}
cosmic_v2_subtype_map("liver")

data("cosmic_v2_sigs")
lihc.pred_cosmic <- predict_exposure(musica = lihc.musica, table_name = "SBS96",
                               signature_res = cosmic_v2_sigs,
                               signatures_to_use =  c(2, 7, 10, 13),
                               algorithm = "lda")
plot_signatures(lihc.pred_cosmic)
plot_exposures(lihc.pred_cosmic)

lihc.pred_our_sigs <- predict_exposure(musica = lihc.musica, table_name = "SBS96",
                                 signature_res = lihc.result, algorithm = "lda")

compare_results(result = lihc.pred_cosmic, other_result = lihc.pred_our_sigs, threshold = 0.40)

```
Annotations:
```{r Annotations}
plot_exposures(lihc.result, group_by = "annotation", 
               annotation = "Tumor_Subtypes", plotly=T)
plot_exposures(lihc.result, proportion = TRUE, 
               group_by = "annotation", annotation = "Tumor_Subtypes")
plot_exposures(lihc.result, plot_type = "box", group_by = "signature", 
               color_by = "annotation", annotation = "Tumor_Subtypes")
```


