---
title: "musicatk_help"
date: "5/16/2021"
output: html_document
---

```{r Import, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Import

#### Import Files

#### Import TCGA Datasets

#### Import Musica


## Create Musica Object


## Annotations
<img src="MusicaTKScreenshots/AnnotationsTab.png" alt="Annotations Tab"/>
<br>

Sample annotations can be used to store information about each sample such as tumor type or treatment status. These are used in downstream plotting functions such as plot_exposures or plot_umap to group or color samples by a particular annotation.

![](MusicaTKScreenshots/AnnotationOptions.png)
<br>
Select the musica object from the "Select object" dropdown, and upload the text file containing sample annotations. After selecting the correct delimiter, a data table will appear below to show the annotations
that will be added to the musica object. Choose the column that contains the sample names from the "Sample Name Columns" dropdown and then click "Add Annotation".
<br>
![](MusicaTKScreenshots/AnnotationTable.png)

## Build Tables
![](MusicaTKScreenshots/TableTab.png)

Generates count tables for different mutation type schemas which can be used as input to the mutational signature discovery or prediction functions. "SBS96" generates a table for single base substitutions following the standard 96 mutation types derived from the trinucleotide context. "SBS192" is the 96 mutation type schema with the addition of transcriptional strand or replication strand information added to each base. "DBS" generates a table for the double base substitution schema used in COSMIC V3. "Indel" generates a table for insertions and deletions following the schema used in COSMIC V3.

To build the count tables, the user must select 1 of the 5 standard motifs in the "Select Count Table" dropdown.

SBS96 - Motifs are the six possible single base pair mutation types times the four possibilities each for upstream and downstream context bases (464 = 96 motifs) <br>
SBS192_Trans - Motifs are an extension of SBS96 multiplied by the transcriptional strand (translated/untranslated), can be specified with "Transcript_Strand". <br>
SBS192_Rep - Motifs are an extension of SBS96 multiplied by the replication strand (leading/lagging), can be specified with "Replication_Strand". <br>
DBS - Motifs are the 78 possible double-base-pair substitutions. <br>
INDEL - Motifs are 83 categories intended to capture different categories of indels based on base-pair change, repeats, or microhomology, insertion or deletion, and length.

In addition to selecting a motif, the user must also provide the reference genome.

## Signatures and Exposures

#### Discover
![](MusicaTKScreenshots/DiscoverTab.png)

Mutational signatures and exposures will be discovered using methods such as Latent Dirichlet Allocation (lda) or Non-Negative Matrix Factorization (nmf). These algorithms will deconvolute a matrix of counts for mutation types in each sample to two matrices: 1) a "signature" matrix containing the probability of each mutation type in each sample and 2) an "exposure" matrix containing the estimated counts for each signature in each sample. Before mutational discovery can be performed, variants from samples first need to be stored in a musica object using the create_musica function and mutation count tables need to be created using functions such as build_standard_table.
<br>
You can select any count table for signature discovery. To obtain biologically significant results, it is important to select a reasonable number of expected signatures. 

#### Predict
![](MusicaTKScreenshots/PredictTab.png)

Exposures for samples will be predicted using an existing set of signatures stored in a musica_result object. Algorithms available for prediction include a modify version of "lda", "decompTumor2Sig", and "deconstructSigs".

![](MusicaTKScreenshots/PredictOptions.png)

The "Signatures to Predict" dropdown contains all the signatures in the result object selected from the 
"Result to Predict" dropdown. You can search this dropdown and select multiple signatures. You can also
remove signatures by deleting text from the select input. To use the "deconstructSigs" algorithm, you 
will need to provide the reference genome.


## Data Visualization


## Additional Analysis

#### Compare Signatures
![](MusicaTKScreenshots/CompareTab.png)

Compare two result objects to find similar signatures. The threshold acts as a cutoff
similarity score and can be any value between 0 and 1. Results will populate in a
data table below which can be downloaded.

![](MusicaTKScreenshots/CompareTable.png)

#### Exposure Differential Analysis
![](MusicaTKScreenshots/DATab.png)
The "Exposure Differential Analysis" tab is used to run differential analysis on the signature exposures of annotated samples. There are 3 methods to perform the differential analysis: Wilcoxon Rank Sum Test, Kruskal-Wallis Rank Sum Test, and a negative binomial regression (glm). 

When using the Wilcoxon Rank Sum Test, any two groups will be compared in a pairwise fashion. Any isolated group will be ignored. Below we display the Wilcoxon Rank Sum Test results between LUAD and LUSC. Note that SKCM was ignored since it has no pair.

![](MusicaTKScreenshots/DAOptions.png)
![](MusicaTKScreenshots/WilcoxonTable.png)


#### Clustering

#### Heatmap


## Download


